src/imex/transfer.rs

```diff
@@ -0,0 +1,514 @@
+//! - The *Getter*, which connects to the server and retrieves the data.
+//!
+//! Sendme is designed around the idea of verifying hashes, the downloads are verified as
+//! they are retrieved.  The entire transfer is initiated by requesting the data of a single
+//! root hash.
+//!
+//! Both the provider and the getter are authenticated:
+//!
+//! - The provider is known by its *peer ID*.
+//! - The provider needs an *authentication token* from the getter before it accepts a
+//!   connection.
+//!
+//! Both these are transferred in the QR code offered to the getter.  This ensures that the
+//! getter can not connect to an impersonated provider and the provider does not offer the
+//! download to an impersonated getter.
+
+use std::path::Path;
+
+use anyhow::{anyhow, bail, ensure, format_err, Context as _, Result};
+use async_channel::Receiver;
+use futures_lite::StreamExt;
+use sendme::get::{DataStream, Options};
+use sendme::progress::ProgressEmitter;
+use sendme::protocol::AuthToken;
+use sendme::provider::{DataSource, Event, Provider, Ticket};
+use sendme::Hash;
+use tokio::fs::{self, File};
+use tokio::io::{self, AsyncWriteExt, BufWriter};
+use tokio::sync::broadcast;
+use tokio::sync::broadcast::error::RecvError;
+use tokio::task::JoinHandle;
+use tokio_stream::wrappers::ReadDirStream;
+
+use crate::chat::delete_and_reset_all_device_msgs;
+use crate::context::Context;
+use crate::qr::Qr;
+use crate::{e2ee, EventType};
+
+use super::{export_database, BlobDirContents, DeleteOnDrop, DBFILE_BACKUP_NAME};
+
+/// Provide or send a backup of this device.
+///
+/// This creates a backup of the current device and starts a service which offers another
+/// device to download this backup.
+///
+/// This does not make a full backup on disk, only the SQLite database is created on disk,
+/// the blobs in the blob directory are not copied.
+///
+/// This starts a task which acquires the global "ongoing" mutex.  If you need to stop the
+/// task use the [`Context::stop_ongoing`] mechanism.
+#[derive(Debug)]
+pub struct BackupProvider {
+    /// The supervisor task, run by [`BackupProvider::watch_provider`].
+    handle: JoinHandle<Result<()>>,
+    /// The ticket to retrieve the backup collection.
+    ticket: Ticket,
+}
+
+impl BackupProvider {
+    /// Prepares for sending a backup to a second device.
+    ///
+    /// Before calling this function all I/O must be stopped so that no changes to the blobs
+    /// or database are happening, this is done by calling the [`Accounts::stop_io`] or
+    /// [`Context::stop_io`] APIs first.
+    ///
+    /// This will acquire the global "ongoing process" mutex.  You must call
+    /// [`BackupProvider::join`] after creating this struct, otherwise this will not respect
+    /// the possible cancellation of the "ongoing process".
+    ///
+    /// [`Accounts::stop_io`]: crate::accounts::Accounts::stop_io
+    pub async fn prepare(context: &Context, dir: &Path) -> Result<Self> {
```

Comment:
Let's start discussing this as well: Here this takes a path to a temporary directory that the caller has to clean up afterwards.  That's not terribly ergonomic for the caller.

The problem is that the core only has the blobdir where it can safely write files, and it needs to write the sqlite export somewhere.  The same problem exist on importing a backup (the sqlite export needs to exist on disk somewhere first) and there I did a, perhaps somewhat dubious, hack: I write the sqlite export to a file in the blobdir, but using an invalid blobname.  This means it can not conflict with any valid blob (important because otherwise you could attack the backup transfer by sending well-named sqlite databases as attachments to your contacts).

We could use the same hack on the provider side as well.  Core would be responsible for cleaning up the file, but that's something it should probably do regardless.

We could also try and formalise this hack somehow and create a permanent temporary directory in the blobdir where core can create and delete files.

/cc @link2xt opinions?